### PR TITLE
Add support for Blocking and Non-blocking put-with-signal

### LINKED ISSUE #206: shmem_put_signal
Discuss a proposal for shmem_put_signal.

Transfers contiguous data from a local data object to a specified processing element (PE) without blocking the caller and sets a remote flag to signal completion.

Cray's current implementation:

```c
void shmem_put_signal[_nb](void *target, const void *source, size_t len, uint64_t *sig_target, uint64_t sig_val, int pe)
```

Semantically (almost) equivalent to :

- `shmem_put[_nb](data)`
- `shmem_fence()`
- `shmem_put[_nb](signal)`

except a `shmem_fence` is 'heavier' than the `shmem_put_signal` which only fences the data before the signal.

- The target buffer and signal location must reside in symmetric memory.
- All data size and type variants are supported.
- signal is always a `uint64_t`.

Attaching slides for RMA WG 03 01 2018
https://github.com/openshmem-org/specification/wiki/RMA-WG-03-01-2018

[OpenSHMEM_RMA_PUTSIGNAL.pptx](https://github.com/openshmem-org/specification/files/1768534/OpenSHMEM_RMA_PUTSIGNAL.pptx)

**Edit** (@nspark): Fixed some Markdown formatting issues.

### PR DESCRIPTION
**Pull request by naveen-rn**
### Proposal went mutliple changes in semantics and documentation after the initial reading - [refer to the following comment](https://github.com/openshmem-org/specification/pull/275#issuecomment-524505062) for the most recent update
Previously, we proposed the support for put-with-signal operation in #206. We worked on separate blocking and non-blocking proposals in #218 and #244. Based on multiple review comments and discussions, we have created a consolidated proposal which includes the following:

1. Support for blocking put-with-signal routines
2. Support for non-blocking put-with-signal routines
3. Library constants to select the type of signal update operation (SET and ADD)
4. Update to point-to-point synchronization routines to return signal value
5. Library constants to provide hints to the implementations on the type of update to be expected on the wait operation

In this proposal, apart from adding support for put-with-signal routines, we have update the point-to-point sync routines - especially ```shmem_wait_until``` to return the values from ```ivar``` on which the wait operation is performed. This would enable us to read the signal update efficiently using the wait operation. The need for the update on the wait operation are provided in the attached
[presentation.](https://github.com/openshmem-org/specification/files/3143910/reading-signal-put-with-signal.pptx)

In general, the proposed put-with-signal has the following atomicity guarantees. 

| signal update type | ```SHMEM_SIGNAL_SET``` | ```SHMEM_SIGNAL_ADD``` | ```shmem_atomics``` | p2p-sync |
|--------------------|-----------------------|------------------------|----------------------|-----------|
| ```SHMEM_SIGNAL_SET``` | exclusive | undefined | undefined | exclusive |
| ```SHMEM_SIGNAL_ADD``` | undefined | exclusive | undefined | exclusive |

The only possible way to read the signal in portable fashion is through the updated ```shmem_wait_until``` operation and passing the hints with the comparison operator would enable us to optimize the wait operation, if provided by the implementation.

Closes #206

### CODE REVIEW

File: content/shmem_put_signal.tex [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    contiguous local data object to a data object on a specified \ac{PE}
+    and subsequently updating a remote flag to signal completion. The routines
+    return after the data has been copied out of the \source{} array on the
+    local \ac{PE}.
+
+    The \VAR{sig\_op} signal operator determines the type of update to be
+    performed on the remote \VAR{sig\_addr} signal data object. The completion
+    of signal update based on the \VAR{sig\_op} signal operator using the
+    \VAR{signal} flag on the remote \ac{PE} indicates the delivery of its
+    corresponding \dest{} data words into the data object on the remote \ac{PE}.
```
Comment (manjugv) [inline content/shmem_put_signal.tex]: May F2F: Add a phrase about sig addr completion.

File: content/shmem_put_signal_nbi.tex [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+void @\FuncDecl{shmem\_ctx\_put\FuncParam{SIZE}\_signal\_nbi}@(shmem_ctx_t ctx, void *dest, const void *source, size_t nelems, uint64_t *sig_addr, uint64_t signal, int sig_op, int pe);
+\end{CsynopsisCol}
+where \SIZE{} is one of \CONST{8, 16, 32, 64, 128}.
+
+\begin{CsynopsisCol}
+void @\FuncDecl{shmem\_putmem\_signal\_nbi}@(void *dest, const void *source, size_t nelems, uint64_t *sig_addr, uint64_t signal, int sig_op, int pe);
+void @\FuncDecl{shmem\_ctx\_putmem\_signal\_nbi}@(shmem_ctx_t ctx, void *dest, const void *source, size_t nelems, uint64_t *sig_addr, uint64_t signal, int sig_op, int pe);
+\end{CsynopsisCol}
+
+\begin{apiarguments}
```
Comment (anshumang) [inline content/shmem_put_signal_nbi.tex]: Sharing a thought, does it make sense to not repeat description text for arguments for the same API (the `_nbi` and the other variant)?

File: content/shmem_put_signal_nbi.tex [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    from a contiguous local data object to a data object on a specified \ac{PE}
+    and subsequently updating a remote flag to signal completion.
+
+    The routines return after posting the operation. The operation is considered
+    complete after a subsequent call to \FUNC{shmem\_quiet}. At the completion
+    of \FUNC{shmem\_quiet}, the data has been copied out of the \source{} array
+    on the local \ac{PE} and delivered into the \dest{} array on the destination
+    \ac{PE}.
+
+    The delivery of \VAR{signal} flag on the remote \ac{PE} indicates only the
```
Comment (anshumang) [inline content/shmem_put_signal_nbi.tex]: This is the only paragraph with text that provides additional info. Wondering if the repitition could be deleted in some way.

**Comment by BryantLam** [comment]
Three options discussed regarding signal operation and interaction with `shmem_wait_until`.
1. Change `SHMEM_SIGNAL_UPDATE` from a hint to a requirement.
2. Introduce new API variants for `wait_until` that work specifically only for the signal operation to preserve performance.
3. Make signal operation atomic compatible with other atomic operations.
Option 2 was selected in a straw poll.

File: main_spec.tex [hunk 4]
```diff
@@ -23,7 +23,7 @@ \section{Execution Model}\label{subsec:execution_model}
 \input{content/execution_model}
 
 \section{Language Bindings and Conformance}\label{subsec:bindings}
-\input{content/language_bindings_and_conformance}
+\input{content/language_bindings_and_conformance}re
```
Comment (nspark) [inline main_spec.tex]: This seems like a stray typo.

File: content/shmem_signal_wait_until.tex [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+\begin{apidefinition}
+
+\begin{Csynopsis}
+uint64_t @\FuncDecl{shmem\_signal\_wait\_until}@(uint64_t *sig_addr, int cmp, uint64_t cmp_value);
+\end{Csynopsis}
+
+\begin{apiarguments}
+
+\apiargument{OUT}{sig\_addr}{A pointer to a remotely accessible variable.}
```
Comment (nspark) [inline content/shmem_signal_wait_until.tex]: 7/22 reading: "IN" not "OUT" argument

**Comment by anshumang** [comment]
Adding a point-to-point synchronization example that use `shmem_put_signal_nbi` would be useful given that this is the only non-blocking API that can be used for point-to-point synchronization.
Use of this API for point-to-point synchronization requires `shmem_quiet` at the initiator that none of the other synchronization APIs so.

**Comment by naveen-rn** [comment]
Discussed with @anshuman offline and decided to add this example as a separate ticket once the main ticket is merged.

File: content/p2p_sync_intro.tex [hunk 6]
```diff
@@ -3,8 +3,8 @@
 object.
 The point-to-point synchronization routines can be used to portably ensure
 that memory access operations observe remote updates in the order enforced by
-the initiator \ac{PE} using the \FUNC{shmem\_fence} and \FUNC{shmem\_quiet}
-routines.
+the initiator \ac{PE} using the put-with-signal, \FUNC{shmem\_fence} and
```
[the remark this answers is not in the document]
Comment (naveen-rn) [reply content/p2p_sync_intro.tex]: Looks like a new PR is required - not needed as part of this proposal

File: content/shmem_signal_wait_until.tex (hunk 5 above)
Comment (naveen-rn) [reply content/shmem_signal_wait_until.tex]: Fixed

File: main_spec.tex (hunk 4 above)
Comment (naveen-rn) [reply main_spec.tex]: fixed

File: content/shmem_put_signal_nbi.tex (hunk 3 above)
Comment (naveen-rn) [reply content/shmem_put_signal_nbi.tex]: This comment is based on old text - things have moved around for the official reading. Closing this comment for now and let the author reopen if this is still an issue.

File: content/shmem_put_signal_nbi.tex (hunk 2 above)
Comment (naveen-rn) [reply content/shmem_put_signal_nbi.tex]: Not sure if I understand the comment correctly - are you saying that the argument description can be combined with the blocking variant? If so, I think that would make the proposal not consistent with the other parts of the text. I'll close the comment for now and fix it in a separate cleanup PR is needed.

File: content/shmem_put_signal.tex (hunk 1 above)
Comment (naveen-rn) [reply content/shmem_put_signal.tex]: fixed

**Comment by naveen-rn** [comment]
## Update for voting in OpenSHMEM Sep 2019 F2F meeting:

### Proposal Overview
In this propoal, we are adding support for signaling operation in OpenSHMEM along with options to read the signal effectively. As part of this proposal, we introduce three signaling operations and a p2p signaling wait operation. We also introduced library constants ```SHMEM_SIGNAL_SET``` and ```SHMEM_SIGNAL_ADD`` to determine the type of signal update type.

#### Signaling Operations
1. blocking put-with-signal (```shmem_signal_put```), 
2. non-blocking put-with-signal (```shmem_signal_put_nbi```), and
3. signaling fetch (```shmem_signal_fetch```).
#### Signaling wait operation
1. ```shmem_signal_wait_until```
#### Atomicity
In general, the proposed put-with-signal has the following atomicity guarantees. 
1. exclusive with itself and with other signaling operation of the same signal update type,
2. exclusive with respect to all p2p sync operation, and
3. undefined with other OpenSHMEM AMO operations.
This semantic allows the most efficient implementation of the signaling operation and extract hardware optimizations.

| signal update type | ```SHMEM_SIGNAL_SET``` | ```SHMEM_SIGNAL_ADD``` | ```shmem_atomics``` | p2p-sync |
|--------------------|-----------------------|------------------------|----------------------|-----------|
| ```SHMEM_SIGNAL_SET``` | exclusive | undefined | undefined | exclusive |
| ```SHMEM_SIGNAL_ADD``` | undefined | exclusive | undefined | exclusive |
#### Reading Signal Update
Because of the above semantics, signal updates can be read only through special routines : ```shmem_signal_fetch``` and ```shmem_signal_wait_until```. While ```shmem_signal_fetch``` is introduced to exclusively read the signal update, ```shmem_signal_wait_until``` can be used to wait on a signal update and return the updated signal value.

### History
We introduced blocking put-with-signal and non-blocking put-with-signal as separate APIs in #218 and #244. In this proposal, we have consolidated those two proposals and incorporated all the suggested changes from the OpenSHMEM specification committee.

### Changes made after July 2019 Telecon Reading

File: content/shmem_quiet.tex [hunk 7]
```diff
@@ -19,9 +19,9 @@
 
 \apidescription{
     The \FUNC{shmem\_quiet} routine ensures completion of \PUT{}, \ac{AMO},
-    memory store, and nonblocking \PUT{} and \GET{} routines on
+    \emph{put-with-signal}, memory store, and nonblocking \PUT{},\emph{put-with-signal}, and \GET{} routines on
```
Comment (naveen-rn) [inline content/shmem_quiet.tex]: The sh_q routine ensure completion of blocking put, amo, pws, and memory store, as well as nonblocking put, pws, and get routines on .....

and -> before memory store
and nonblocking -> as well as nonblocking

File: content/shmem_fence.tex [hunk 8]
```diff
@@ -17,12 +19,18 @@
 \end{apiarguments}
 
 \apidescription{
-    This routine assures ordering of delivery of \PUT{}, \ac{AMO}, memory store, and nonblocking \PUT{}
-    routines to symmetric data objects.  All \PUT{}, \ac{AMO}, memory store, and nonblocking \PUT{}
+    This routine assures ordering of delivery of memory store, blocking \PUT{},
+    \ac{AMO}, and \emph{put-with-signal}, as well as nonblocking \PUT{},
+    and \emph{put-with-signal}
+    routines to symmetric data objects.  All memory store, blocking \PUT{},
+    \ac{AMO}, amd \emph{put-with-signal}, as well as nonblocking \PUT{},
```
Comment (jdinan) [inline content/shmem_fence.tex]:
```suggestion
    \ac{AMO}, and \emph{put-with-signal}, as well as nonblocking \PUT{},
```

File: content/shmem_fence.tex [hunk 9]
```diff
@@ -1,6 +1,8 @@
 \apisummary{
-    Assures ordering of delivery of \PUT{}, \ac{AMO}, memory store, and nonblocking \PUT{} routines
-    to symmetric data objects.
+    Assures ordering of delivery of memory store, blocking \PUT{}, 
+    \ac{AMO}, and \emph{put-with-signal}, as well as nonblocking 
+    \PUT{}, and \emph{put-with-signal}
```
Comment (davidozog) [inline content/shmem_fence.tex]:
```suggestion
    \PUT{} and \emph{put-with-signal} 
```

File: content/shmem_fence.tex (hunk 8 above)
Comment (davidozog) [inline content/shmem_fence.tex]:
```suggestion
    \ac{AMO}, amd \emph{put-with-signal}, as well as nonblocking \PUT{}
```

File: content/shmem_fence.tex [hunk 10]
```diff
@@ -17,12 +19,18 @@
...
+    and \emph{put-with-signal}
+    routines to symmetric data objects.  All memory store, blocking \PUT{},
+    \ac{AMO}, amd \emph{put-with-signal}, as well as nonblocking \PUT{},
+    and \emph{put-with-signal}
     routines to symmetric data objects issued to a particular remote \ac{PE}
     on the given context prior
     to the call to \FUNC{shmem\_fence} are guaranteed to be delivered before any
-    subsequent \PUT{}, \ac{AMO}, memory store, and nonblocking \PUT{} routines to symmetric data
+    subsequent memory store, blocking \PUT{}, \ac{AMO}, and \emph{put-with-signal}, 
+    as well as nonblocking \PUT{}, and \emph{put-with-signal}
```
Comment (davidozog) [inline content/shmem_fence.tex]:
```suggestion
    as well as nonblocking \PUT{} and \emph{put-with-signal}
```

File: content/shmem_fence.tex [hunk 11]
```diff
@@ -42,17 +50,19 @@
     ordering of its memory accesses.
     There is a subtle difference between
     \FUNC{shmem\_fence} and \FUNC{shmem\_quiet}, in that, \FUNC{shmem\_quiet}
-    guarantees completion of \PUT{}, \ac{AMO}, memory store, and nonblocking \PUT{} routines to
+    guarantees completion of memory store, blocking \PUT{}, \ac{AMO}, and
+    \emph{put-with-signal}, as well as nonblocking \PUT{}, and \emph{put-with-signal} routines to
```
Comment (davidozog) [inline content/shmem_fence.tex]:
```suggestion
    \emph{put-with-signal}, as well as nonblocking \PUT{} and \emph{put-with-signal} routines to
```

File: content/shmem_fence.tex [hunk 12]
```diff
@@ -42,17 +50,19 @@
...
     symmetric data objects which makes the updates visible to all other
     \acp{PE}.
 
-    The \FUNC{shmem\_quiet} routine should be called if completion of \PUT{},
-    \ac{AMO}, memory store, and nonblocking \PUT{} routines to symmetric data objects is desired
+    The \FUNC{shmem\_quiet} routine should be called if completion of memory
+    store, blocking \PUT{}, \ac{AMO}, and \emph{put-with-signal}, as well as
+    nonblocking \PUT{}, and \emph{put-with-signal} routines to symmetric data objects is desired
```
Comment (davidozog) [inline content/shmem_fence.tex]:
```suggestion
    nonblocking \PUT{} and \emph{put-with-signal} routines to symmetric data objects is desired
```

**Pull request merged**